### PR TITLE
Pose, twist and odometry with covariance msgs

### PR DESCRIPTION
**Pull request by adityapande-1995**
Signed-off-by: Aditya <aditya050995@gmail.com>

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

# 🎉 New feature

## Summary
This PR adds 2 new message types : ``TwistWithCovariance`` and ``PoseWithCovariance`` which have covariance matrices along with the original twist / pose message. This type was available in ``geometry_msgs/*WithCovariance`` and was used in ``gazebo_ros_pkgs`` in cases where the poses have noise associated with them.

Also fixes a minor typo in ``twist.proto``

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

**Review by scpeters — APPROVED** [review]

**Comment by scpeters** [comment]
actually, have you tested whether the ros_ign bridge properly converts this to the corresponding ROS message?

**Comment by scpeters** [comment]
I'd rather not merge this to a stable branch if it hasn't been tested with the bridge

**Comment by adityapande-1995** [comment]
I haven't tested it with the ``ros_ign`` bridge yet. On it :+1:

File: proto/ignition/msgs/odometry_with_covariance.proto [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
```
Comment (mjcarroll) [inline proto/ignition/msgs/odometry_with_covariance.proto:2]:
```suggestion
 * Copyright (C) 2022 Open Source Robotics Foundation
```

**Review by mjcarroll — COMMENTED** [review]
One tiny nit

File: proto/ignition/msgs/pose_with_covariance.proto [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+message PoseWithCovariance
+{
+  /// \brief Pose message.
+  Pose pose          = 1;
+
+  /// \brief Row-major representation of the 6x6 covariance matrix
+  /// The orientation parameters use a fixed-axis representation.
+  /// In order, the parameters are:
+  /// (x, y, z, rotation about X axis, rotation about Y axis, rotation about Z axis)
+  Float_V covariance          = 2;
```
Comment (jacobperron) [inline proto/ignition/msgs/pose_with_covariance.proto:39]: All the space before the `=` looks strange to me, I guess it's intended for alignment with other fields? If so maybe we should remove some space from this line so the `=` is in the same column as the other `=` on L33?

File: proto/ignition/msgs/twist_with_covariance.proto [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+message TwistWithCovariance
+{
+  /// \brief Twist message.
+  Twist twist          = 1;
+
+  /// \brief Row-major representation of the 6x6 covariance matrix
+  /// The orientation parameters use a fixed-axis representation.
+  /// In order, the parameters are:
+  /// (x, y, z, rotation about X axis, rotation about Y axis, rotation about Z axis)
+  Float_V covariance          = 2;
```
Comment (jacobperron) [inline proto/ignition/msgs/twist_with_covariance.proto:39]: Same here.

**Review by jacobperron — APPROVED** [review]
LGTM, one question about style below.

File: proto/ignition/msgs/odometry_with_covariance.proto (hunk 1 above)
Comment (adityapande-1995) [reply proto/ignition/msgs/odometry_with_covariance.proto]: Done

File: proto/ignition/msgs/pose_with_covariance.proto (hunk 2 above)
Comment (adityapande-1995) [reply proto/ignition/msgs/pose_with_covariance.proto]: Done

File: proto/ignition/msgs/twist_with_covariance.proto (hunk 3 above)
Comment (adityapande-1995) [reply proto/ignition/msgs/twist_with_covariance.proto]: Done

File: proto/ignition/msgs/odometry_with_covariance.proto [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+message OdometryWithCovariance
+{
+  /// \brief Optional header data
+  Header header                             = 1;
+
+  /// \brief Estimated pose.
+  PoseWithCovariance pose_with_covariance   = 2;
+
+  /// \brief Estimated velocity.
```
Comment (caguero) [inline proto/ignition/msgs/odometry_with_covariance.proto:38]: Is this comment about velocity correct?

**Comment by adityapande-1995** [comment]
PR to bridge these msgs to ROS : https://github.com/ignitionrobotics/ros_ign/pull/222

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-25-fortress-edifice-citadel/1343/1

**Pull request merged**